core/market/src/api/provider.rs

```diff
@@ -0,0 +1,222 @@
+use super::{
+    resolve_web_error, PathAgreement, PathSubscription, PathSubscriptionProposal, QueryTimeout,
+    QueryTimeoutMaxEvents,
+};
+use crate::utils::*;
+use actix_web::web::{Data, Json, Path, Query};
+use actix_web::HttpResponse;
+use std::time::Duration;
+use ya_client::{
+    market::MarketProviderApi,
+    web::{WebAuth, WebClient},
+    Result,
+};
+use ya_client_model::market::*;
+use ya_persistence::executor::DbExecutor;
+use ya_service_api_web::middleware::Identity;
+
+pub fn extend_web_scope(scope: actix_web::Scope) -> actix_web::Scope {
+    scope
+        .service(subscribe)
+        .service(get_offers)
+        .service(unsubscribe)
+        .service(collect)
+        .service(counter_proposal)
+        .service(get_proposal)
+        .service(reject_proposal)
+        .service(approve_agreement)
+        .service(reject_agreement)
+}
+
+// ****************************************************
+
+fn build_market_api(id: Identity) -> Result<MarketProviderApi> {
+    let client_result = WebClient::builder()
```

Comment:
we need to reuse Actix Web Clients as much as possible

Reply:
Please suggest a way to reuse WebClient instances in a way that would be thread safe in Rust. I am guessing that some sort of a global WebClient cache would be required (as there is a need for one WebClient per each Identity calling), though trying to implement it via HashMap yields:

51 | static proxy_cache: HashMap<NodeId, MarketRequestorApi> = HashMap::new();
   |                                                           ^^^^^^^^^^^^^^

error[E0277]: `std::rc::Rc<url::Url>` cannot be shared between threads safely

So I'm all open for suggestions as to how to implement this properly.

Reply:
All api methods get DbExecutor as first parametr. You could replace it with your own struct that stores web client (wrapped in Arc od course). Check activity api how to do this. I write from phone and can't help more. I would be against static cache.

Reply:
I'm on it

Reply:
just noticed your comments guys. I'm working on this
1. we do not need to remove DbExecutor because [`actix_web::Scope.data`](https://docs.rs/actix-web/2.0.0/actix_web/struct.Scope.html#method.data) is indexed by input type id. Currently as far as I can see we do not use it anywhere. It might be useful for cacheing `get_agreement` in local DB as it is in current GSB API call. But now I see it might cause inconsistency problems because Agreement can change state...
1. [`actix_web::web::Data`](https://docs.rs/actix-web/2.0.0/actix_web/web/struct.Data.html) internally uses `Arc` already, so there is no need to use any reference counting